File: libsio/torchscript.cc
#include <torch/torch.h>
#include <torch/script.h>

int main() {
    torch::jit::script::Module module;
    module = torch::jit::load("model.pt");
    // Create a vector of inputs.
    std::vector<torch::jit::IValue> inputs;
    inputs.push_back(torch::ones({10, 20}));

    // Execute the model and turn its output into a tensor.
    torch::Tensor output1 = module.forward(inputs).toTensor();
    std::cout << output1 << "\n";

    torch::Tensor input_a = torch::ones({10, 20});
    torch::Tensor input_b = torch::ones({10, 20});
    torch::Tensor output2 = module.run_method("another_forward", input_a, input_b).toTensor();
    std::cout << output2 << "\n";

    return 0;
}

File: libsio/torchscript.py
import torch

class MyModule(torch.nn.Module):
    def __init__(self, N, M):
        super(MyModule, self).__init__()
        self.weight = torch.nn.Parameter(torch.ones(N, M))

    def forward(self, input):
        if input.sum() > 0:
          output = self.weight + input
        else:
          output = -self.weight
        return output

    @torch.jit.export
    def another_forward(self, input_a, input_b):
        if (input_a + input_b).sum() > 0:
          output = self.weight - 1
        else:
          output = -self.weight
        return output

my_module = MyModule(10,20)
sm = torch.jit.script(my_module)

print(sm.graph)
print(sm.code)

sm.save("model.pt")

